zipline/utils/sentinel.py

```diff
@@ -4,22 +4,42 @@
 Sentinel objects are used when you only care to check for object identity.
 """
 import sys
+from textwrap import dedent
 
 
 def sentinel(name, doc=None):
-    @object.__new__   # bind a single instance to the name 'NotSpecified'
-    class result(object):
+    try:
+        value = sentinel._cache[name]  # memoized
+    except KeyError:
+        pass
+    else:
+        if doc == value.__doc__:
+            return value
+
+        raise ValueError(dedent(
+            """\
+            New sentinel value %r conflicts with an existing sentinel of the
+            same name.
+            Old sentinel docstring: %r
+            New sentinel docstring: %r
+            Resolve this conflict by changing the name of one of the sentinels.
+            """,
+        ) % (name, value.__doc__, doc))
+
+    @object.__new__   # bind a single instance to the name 'Sentinel'
+    class Sentinel(object):
         __doc__ = doc
         __slots__ = ('__weakref__',)
+        __name__ = name
 
         def __new__(cls):
-            raise TypeError("Can't construct new instances of %s" % name)
+            raise TypeError("Can't construct new instances of %r" % name)
```

Comment:
I think I'd counter that we should make our error messages complete sentences.  Grammar and punctuation exist because they help convey meaning.  The fact that we're writing code doesn't make that suddenly not true.